### PR TITLE
feat: tracing Random.jl functionality correctly

### LINKED ISSUE #356: Handling random numbers correctly
```julia
julia> function test_rand(x)
           y = similar(x)
           randn!(y)
           return y
       end
test_rand (generic function with 1 method)

julia> test_rand(rand(2, 3))
2×3 Matrix{Float64}:
 -0.856593   2.09777  -1.47022
 -0.587353  -1.63428   0.101531

julia> test_rand(rand(2, 3))
2×3 Matrix{Float64}:
 -0.169647  0.20255    1.73269
  1.99007   0.0300865  1.00466

julia> @code_hlo test_rand(Reactant.to_rarray(rand(2, 3)))
module {
  func.func @main(%arg0: tensor<3x2xf64>) -> tensor<3x2xf64> {
    %cst = stablehlo.constant dense<[[-0.14388180121311803, -1.7968975324220513], [0.40886485616904072, -0.26052122391271643], [-2.1956924346267952, -0.047720339420302453]]> : tensor<3x2xf64>
    return %cst : tensor<3x2xf64>
  }
}

julia> @code_hlo test_rand(Reactant.to_rarray(rand(2, 3)))
module {
  func.func @main(%arg0: tensor<3x2xf64>) -> tensor<3x2xf64> {
    %cst = stablehlo.constant dense<[[-0.39078327584480177, 0.46390776775279169], [-0.2758724528591438, -2.4972617545675377], [0.51502111596476896, -0.56535270904132628]]> : tensor<3x2xf64>
    return %cst : tensor<3x2xf64>
  }
}

julia> fn = @compile test_rand(Reactant.to_rarray(rand(2, 3)))
Reactant.Compiler.Thunk{Symbol("##test_rand_reactant#655815")}()

julia> fn(Reactant.to_rarray(rand(2, 3)))
2×3 ConcreteRArray{Float64, 2}:
 -0.788288  -1.67923   0.00359754
  0.268338   1.13756  -0.907077

julia> fn(Reactant.to_rarray(rand(2, 3)))
2×3 ConcreteRArray{Float64, 2}:
 -0.788288  -1.67923   0.00359754
  0.268338   1.13756  -0.907077
```

This is not unexpected from how we do the tracing but definitely not the preferred behavior

### PR DESCRIPTION
**Pull request by avik-pal**
## TODOs

- [x] Tests
- [ ] ~Overlay seems to act weird. Can't switch to non-reactant interpreter nicely~ Using a workaround for now
- [x] Fix documentation build
- [x] Overlay `Random123` to generate specific RNGs (Threefry and Philox)
- [x] Overlay all random number generators
  - [x] construct TracedRNG during tracing

### CODE REVIEW

**Comment by avik-pal** [comment]
```julia
julia> using Reactant, Random

julia> fn() = randn(Random.default_rng(), 2, 3)
fn (generic function with 1 method)

julia> @code_hlo optimize = false fn()
module {
  func.func @main() -> tensor<3x2xf64> {
    %c = stablehlo.constant dense<[14557082219554432694, 7679360750247989405]> : tensor<2xui64>
    %output_state, %output = stablehlo.rng_bit_generator %c, algorithm =  DEFAULT : (tensor<2xui64>) -> (tensor<2xui64>, tensor<2x3xf64>)
    %cst = stablehlo.constant dense<2.000000e+00> : tensor<2x3xf64>
    %0 = stablehlo.multiply %output, %cst : tensor<2x3xf64>
    %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<2x3xf64>
    %1 = stablehlo.subtract %0, %cst_0 : tensor<2x3xf64>
    %2 = chlo.erf_inv %1 : tensor<2x3xf64> -> tensor<2x3xf64>
    %cst_1 = stablehlo.constant dense<1.4142135623730951> : tensor<2x3xf64>
    %3 = stablehlo.multiply %2, %cst_1 : tensor<2x3xf64>
    %4 = stablehlo.transpose %3, dims = [1, 0] : (tensor<2x3xf64>) -> tensor<3x2xf64>
    return %4 : tensor<3x2xf64>
  }
}

julia> @code_hlo optimize = true fn()
module {
  func.func @main() -> tensor<3x2xf64> {
    %cst = stablehlo.constant dense<1.4142135623730951> : tensor<2x3xf64>
    %cst_0 = stablehlo.constant dense<1.000000e+00> : tensor<2x3xf64>
    %cst_1 = stablehlo.constant dense<2.000000e+00> : tensor<2x3xf64>
    %c = stablehlo.constant dense<[11413925565016077410, 2794483660876742882]> : tensor<2xui64>
    %output_state, %output = stablehlo.rng_bit_generator %c, algorithm =  DEFAULT : (tensor<2xui64>) -> (tensor<2xui64>, tensor<2x3xf64>)
    %0 = stablehlo.multiply %output, %cst_1 : tensor<2x3xf64>
    %1 = stablehlo.subtract %0, %cst_0 : tensor<2x3xf64>
    %2 = chlo.erf_inv %1 : tensor<2x3xf64> -> tensor<2x3xf64>
    %3 = stablehlo.multiply %2, %cst : tensor<2x3xf64>
    %4 = stablehlo.transpose %3, dims = [1, 0] : (tensor<2x3xf64>) -> tensor<3x2xf64>
    return %4 : tensor<3x2xf64>
  }
}
```

**Comment by avik-pal** [comment]
This is kind of working now. Can I get an initial review?

File: src/Interpreter.jl [hunk 1]
```diff
@@ -98,6 +98,18 @@ function set_reactant_abi(
...
+    # https://github.com/EnzymeAD/Reactant.jl/issues/356
+    if (f === Random.default_rng || f === default_rng) && length(argtypes) == 1
+        arginfo2 = ArgInfo(
+            fargs isa Nothing ? nothing : Any[:($(default_rng_inside_interpreter))],
+            Any[Core.Const(default_rng_inside_interpreter)],
+        )
+        return abstract_call_known(
+            interp, default_rng_inside_interpreter, arginfo2, si, sv, max_methods
+        )
+    end
```
Comment (avik-pal) [inline src/Interpreter.jl:111]: This will have to be updated once the new CUDA interpreter stuff lands

**Review by wsmoses — COMMENTED** [review]
This looks good to me but should wait until the interpreter lands (which will simplify the override and can just use the macro)

File: test/ops.jl [hunk 2]
```diff
@@ -538,8 +538,34 @@ end
 end
 
 @testset "rng_bit_generator" begin
-    # seed = ConcreteRArray([0, 0])
-    # @jit Ops.rng_bit_generator(seed, [2])
+    genInt32(seed) = Ops.rng_bit_generator(Int32, seed, [2, 4])
+    genInt64(seed) = Ops.rng_bit_generator(Int64, seed, [2, 4])
+    genUInt64(seed) = Ops.rng_bit_generator(UInt64, seed, [2, 4])
```
Comment (mofeing) [inline test/ops.jl:543]: maybe also `genFloat64`?

**Comment by avik-pal** [comment]
The overlay mechanism fixed the previous unreachable instruction issue :tada:

File: src/Ops.jl [hunk 3]
```diff
@@ -1017,18 +1017,46 @@ end
...
+    )
     return (;
-        output_state=TracedRArray{UInt64,1}((), MLIR.IR.result(op, 1), MLIR.IR.size(seed)),
-        output=TracedRArray{T,length(shape)}((), MLIR.IR.result(op, 2), shape),
+        output_state=TracedRArray{UInt64,1}((), MLIR.IR.result(op, 1), size(seed)),
+        output=TracedRArray{T,length(shape)}((), MLIR.IR.result(op, 2), Tuple(shape)),
+    )
+end
+
+function rng_bit_generator(
```
Comment (wsmoses) [inline src/Ops.jl:1045]: noinline please!

File: src/Overlay.jl [hunk 4]
```diff
@@ -22,3 +31,60 @@ end
...
+        end
+
+        # warn about direct writing to arrays
+        @reactant_overlay @noinline function Random.$(randfun!)(
+            rng::AbstractRNG, A::AbstractArray
+        )
+            @warn "Directly writing to an array using Random.jl functions inside \
+                   ReactantInterpreter will generate a constant array in the IR. Use with \
+                   caution." maxlog = 1
+            return Random.$(randfun!)(rng, A)
```
Comment (avik-pal) [inline src/Overlay.jl:87]: @wsmoses my understanding was that this should call the non-overlayed version. But here I get 

```
┌ Warning: Directly writing to an array using Random.jl functions inside ReactantInterpreter will generate a constant array in the IR. Use with caution.
└ @ Reactant /mnt/software/lux/Reactant.jl/src/Overlay.jl:84
Unreachable reached at 0x7a64877b6b16

[1417783] signal 4 (2): Illegal instruction
in expression starting at REPL[7]:1
fn2 at ./REPL[6]:4 [inlined]
opaque closure at ./<missing>:0
unknown function (ip: 0x7a64877b6bff)
fn2 at ./REPL[6]:2 [inlined]
call_with_reactant at /mnt/software/lux/Reactant.jl/src/utils.jl:0
#8 at /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:210
block! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
unknown function (ip: 0x7a64877b6316)
#make_mlir_fn#1 at /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:197
make_mlir_fn at /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:117 [inlined]
#10 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:295 [inlined]
block! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
#9 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:294 [inlined]
mmodule! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Module.jl:92
unknown function (ip: 0x7a64877b5976)
#compile_mlir!#8 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:291
compile_mlir! at /mnt/software/lux/Reactant.jl/src/Compiler.jl:290 [inlined]
#6 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:285 [inlined]
context! at /mnt/software/lux/Reactant.jl/src/mlir/IR/Context.jl:76
unknown function (ip: 0x7a64877b2a76)
#compile_mlir#5 at /mnt/software/lux/Reactant.jl/src/Compiler.jl:283
compile_mlir at /mnt/software/lux/Reactant.jl/src/Compiler.jl:280
unknown function (ip: 0x7a64877b0a66)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
do_call at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:126
eval_value at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:223
eval_stmt_value at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:174 [inlined]
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:663
jl_interpret_toplevel_thunk at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:821
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:943
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:625
eval_body at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:539
jl_interpret_toplevel_thunk at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/interpreter.c:821
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:943
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
jl_toplevel_eval_flex at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:886
ijl_toplevel_eval_in at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/toplevel.c:994
eval at ./boot.jl:430 [inlined]
eval_user_input at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:245
repl_backend_loop at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:342
#start_repl_backend#59 at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:327
start_repl_backend at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:324
#run_repl#72 at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:483
run_repl at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/share/julia/stdlib/v1.11/REPL/src/REPL.jl:469
jfptr_run_repl_10705 at /mnt/.julia/compiled/v1.11/REPL/u0gqU_FGbh7.so (unknown line)
#1150 at ./client.jl:446
jfptr_YY.1150_15174 at /mnt/.julia/compiled/v1.11/REPL/u0gqU_FGbh7.so (unknown line)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
jl_f__call_latest at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/builtins.c:875
#invokelatest#2 at ./essentials.jl:1055 [inlined]
invokelatest at ./essentials.jl:1052 [inlined]
run_main_repl at ./client.jl:430
repl_main at ./client.jl:567 [inlined]
_start at ./client.jl:541
jfptr__start_73406.1 at /home/avikpal/.julia/juliaup/julia-1.11.2+0.x64.linux.gnu/lib/julia/sys.so (unknown line)
jl_apply at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/julia.h:2157 [inlined]
true_main at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/jlapi.c:900
jl_repl_entrypoint at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/src/jlapi.c:1059
main at /cache/build/tester-amdci5-12/julialang/julia-release-1-dot-11/cli/loader_exe.c:58
unknown function (ip: 0x7a65341e2e07)
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
unknown function (ip: 0x4010b8)
Allocations: 48616895 (Pool: 48615481; Big: 1414); GC: 47
[1]    1417783 illegal hardware instruction (core dumped)  julia --project=envs --threads=4 --check-bounds=yes
```

File: src/stdlibs/Random.jl [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+# call from overlay-ed variants. we write this with 2 tiers -- overload_* and
+# internal_overload_* -- to avoid method ambiguities
+for randfun in (:rand, :randn, :randexp, :rand!, :randn!, :randexp!)
+    overload_randfun = Symbol(:overload_, randfun)
+    internal_overload_randfun = Symbol(:internal_overload_, randfun)
+    @eval begin
+        function $(overload_randfun)(rng::AbstractRNG, args...)
+            # XXX: Ideally the following should just work but currently it gives an illegal
+            #      instruction error. Maybe an issue with Julia's AbsInt?
+            # seed_uint64 = rand(rng, UInt64, 2)
```
Comment (avik-pal) [inline src/stdlibs/Random.jl:138]: same problem here. calling this leads to the illegal instruction

File: src/Overlay.jl (hunk 4 above)
Comment (avik-pal) [reply src/Overlay.jl]: This is definitely a general issue with any kind of recursion of an overlayed function. Do we have a way to force using the NativeInterpreter?

Comment (wsmoses) [reply src/Overlay.jl]: ccing @gbaraldi and @aviatesk

**Comment by avik-pal** [comment]
I am working around the AbsInt issues for now, but those would be nice to sort out, especially for handling the wrapped arrays.

Also, I am explicitly throwing errors in the cases where otherwise Julia would crash with an invalid instruction

**Pull request merged**